Sites: Wait for site counts to match before redirecting to all sites

ryelle opened the pull request:
See #5571 — After connecting a Jetpack site, the `me/sites` endpoint can have a delay before registering that the user has a new site. For users, this means we sometimes drop them into the all sites view instead of the view for the site they just connected.

Since the user's site count is correct, this PR changes the site selection code to wait until the site count and number of sites returned is equal. This doesn't address the underlying API cause, it's more of a workaround.

I dropped a Drake "error" on the page to give the user some feedback, but I'd like real designer feedback if we go this route :)

To test:

1. Have an unconnected Jetpack site
2. Clear localStorage and IndexedDB (possibly only necessary if you've connected this site in the past?)
3. Connect your Jetpack site
4. Immediately try to view any calypso page for that site: `/plugins/$site`, `/pages/$site`, etc
5. You should see a "wait" message with Drake, then the page you're expecting (as opposed to the all sites view)

cc @roccotripaldi @rickybanister -- I left this as "In Progress" since I'd like to have some discussion about whether we think this is too much "workaround a problem" and not enough "solve the problem" :)

Test live: https://calypso.live/?branch=add/wait-for-sites-list

mtias commented on client/my-sites/controller.js:
```diff
@@ -183,6 +185,18 @@ module.exports = {
+						</Main>
+					), document.getElementById( 'primary' ) );
+
+					setTimeout( () => {
```
Why not just dispatch a sites fetch action here and wait for it to resolve instead of the timeout?

mtias replied:
I'd model it on the `sites.once` call here: https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/controller.js#L175-L190

- You check if site_count is different from sites.length.
- If so you send a sites fetch and listen to the change event.
- Run the comparison again to know you got the site.

mtias replied:
Basically the logic in that place would be "once sites resolve, if site length is still not the same as the site_count attribute of the user, run another fetch and wait for the change event again".

johnHackworth replied:
We could do this without using timeouts, but relying 100% in events, if we extract the inline callback used on sites.once to its own function. Something line this, replacing lines 184-203:
```
const selectOnSitesChange = () => {
    // if sites have loaded, but siteID is invalid, redirect to allSitesPath                        
    if ( sites.select( siteID ) ) {
        onSelectedSiteAvailable();
    } else if ( ( currentUser.visible_site_count !== sites.getVisible().length ) ) {
        ReactDom.render( (
            <Main>
                <JetpackManageErrorPage template="waitForJetpackSite" />
            </Main>
        ), document.getElementById( 'primary' ) );

        sites.once( 'change', selectOnSitesChange );
        // We haven't set the main site, so we don't want to continue `next()`
        return;
    } else {
        page.redirect( allSitesPath );
    }
}

sites.once( 'change', selectOnSitesChange );
```

ryelle replied:
I started down that path almost exactly at first, then scrapped it for some reason 😔  

Oh -- I think immediately re-pinging for sites was sending more API requests before success than the artificial 1 second delay (it was about 5 to 1). If that's OK, I'll bring back this approach.

roccotripaldi commented:
Thanks so much for getting a solution out there. I think this is on the right track.

While it does work (!!), the Drake screen is a little spotty:
https://cloudup.com/cEuTfm7OrZS

From a philosophical standpoint, I think this is a great solution, provided we can get it to be a bit smoother. A "work around" solution is exactly what we need. Eventually, when we are sending all Jetpack connections through Calypso, this issue should theoretically disappear. For the time being, we need a band-aid to help with NUX.

ryelle commented:
@roccotripaldi thanks for the screen recording - I bet that's because the page is reloading. I'm going to update the PR to what @johnHackworth said above, that should fix the flashing Drake.

ryelle commented:
Updated the PR with the events-based approach. I had to remove the Drake error, since it was replacing the components already rendered in the primary container. Instead, I've added an info notice to let users know to be patient (exact words to be determined 🙂 ).

![screen shot 2016-08-03 at 5 08 02 pm](https://cloud.githubusercontent.com/assets/541093/17382163/7c50a720-599d-11e6-8352-2d52ccd5cc38.png)

It goes away when the site loads.

roccotripaldi commented:
On the posts page for a newly connected site, I got the 'Waiting for Site' notice.
And after a few seconds, the posts loaded, but something went wrong. Every thing froze, and the notice did not disappear.
Got this in the console: 
`post.jsx:249 Uncaught TypeError: Cannot read property 'single_user_site' of undefined`

<img width="951" alt="screen shot 2016-08-04 at 11 00 04 am" src="https://cloud.githubusercontent.com/assets/2694219/17406792/ec9f02c0-5a32-11e6-8ca8-7318429fbf36.png">

roccotripaldi commented:
However, it worked smoothly on the plugins page

roccotripaldi commented:
Just trying out some other key pages, such as the General Settings page. Like the posts page, it started out smooth, but crashed with this:
`controller.js:131 Uncaught (in promise) TypeError: site.fetchSettings is not a function(…)`

ryelle commented:
😞  I bet it's from trying to continue loading the page without a set `site` yet. I'll see if there's anything we can do about that.

ryelle commented:
Okay... 3c06906 somewhat fixes the issue, but only on some pages. General Settings, for example, still has that error. I'm having trouble replicating this consistently, so I haven't been able to test a lot of other pages, but I wouldn't be surprised if other pages follow Settings' approach.

I think we either need to check that _every page_ checks if `site` exists before calling anything on it, or figure out some way to pause the page render until we have the site.

ryelle commented:
What if we redirected to a new animated "please wait" kind of page, then redirected back when the site loads? (might be a terrible idea, I just don't want to have to audit every single calypso page 😱 )

roccotripaldi commented:
> What if we redirected to a new animated "please wait" kind of page

Hmm. That could be helpful. The page could be void of any components that depend on anything sites related. In theory, it seems like a sound idea.
@mtias, can you chime in?

rickybanister commented:
Redirecting to some kind of waiting page seems like a bummer. 

I'm not going to pretend to understand all of the technical stuff because a lot of really smart people have been working toward a solution, but is there anything else we can do, get creative?

Like, the moment they go to connect or authorize, can we start building the site list before we need it? Some kind of pre-emptive thing? Could we work up some kind of full-screen modal guided tour type step? Could we take a cue from the icon-soup-animation from the .com onboarding process (I realize that's the same as a please wait page, but at least it animates)?

ryelle commented:
>  Could we take a cue from the icon-soup-animation from the .com onboarding process (I realize that's the same as a please wait page, but at least it animates)?

That's exactly the one I was thinking of when I said "animated please wait" page.

As for the other ideas - I'm not totally familiar with what connect does, but the issue is that the site isn't in the API yet, so we'd have them waiting for a while in the connection flow, I think. A full screen guide could work, but they're likely only on the page for 2ish seconds, not long enough to really read something.

johnHackworth commented:
> Like, the moment they go to connect or authorize, can we start building the site list before we need it? Some kind of pre-emptive thing?

this is already the case. Those users who have connected their site using the new connect flow won't have this problem at all, because their calypso sites list is updated during the authorization process.

johnHackworth commented:
So the problem is that there are part of the code where we are expecting to have a site, but they are not using the sites middleware, isn't it? if this is the case, we probably should discuss it more broadly within calypso slack channel, because I think that it's a problem that shouldn't exist

johnHackworth commented:
oh wait, I suspect that if we add a `return;` after https://github.com/Automattic/wp-calypso/pull/7250/files#diff-f11e4a0f3fa89aa5176d3244e323a240R195, everything would work . If we don't let that `next()` at the end of the method to be executed, the load of the rest of the middlewares will stop and the site won't be able to crash due site var not being defined

ryelle commented:
That would only return from the callback, not the whole function. Maybe if we move the `next()` call into the callback, so it doesn't run until we have the site... but then I think we bypass the user feedback of placeholder components.

I'm not sure what you mean by "sites middleware" in the previous comment. Most components assume that if `sites.initialized` is true, we have all the sites, or that if the change event fires, then the sites finished loading. Neither of those are true now, since change can fire multiple times, and initialized is set on the first response from the API.

I don't think it's wrong for sites to use `sites.initialized`, which is why my approach to 3c06906 is to override that property - but some other sections use the change event. So if we were going to start a discussion it would probably be about how to check when the site is available.

johnHackworth commented:
This is what I mean, take a look:

https://github.com/Automattic/wp-calypso/blob/master/client/my-sites/site-settings/index.js#L31

When you go to wp.com/settings/general/site_slug , first the siteSelection middleware is executed, and once it's done, it keeps executing the rest of the middlewares that keep executing the rest of the page. 

At first (in my first comment) I assumed that some controller method ( in the case of general settings, that should be `settingsController.siteSettings`) was trying to use `site` without running the siteSelection as a middleware first (since siteSelection should make sure that you can't run a method that requires a selected site without a real site selected), but what happens is that with the new additions we are refetching the sites, but we are also executing always the "next()" call that allows the next middleware associated to that route to be executed (so we end executing .siteSettings before the siteslist is refreshed, and before there's a real 'selectedSite' object).

You are right about the `return;`, thought. I think we should add that return after `201: sites.once ` so the next() is not executed when we are still waiting for the sites to be re-fetched, and another "next()" execution in line 187 so the controller keeps executing the rest of the middlewares once the sites are refetched.  But take what I'm saying with a pinch of salt, because my brain is working at 50% right now :D 

I think i'm doing a pretty awful job trying to explain what I mean, sorry :/

ryelle commented:
@johnHackworth I understood what you meant, and I think it's working! 😀 

I've tested on a few sites (recently connected, existing, non-existent) and it seems to work as expected. There's a slight delay before we give the user feedback on the recently connected sites, as nothing shows up while we're waiting on the sites list refresh (but they actually get to their site at the end of it, so, better experience than the all-sites redirect 👍 ).

johnHackworth commented:
yay! it looks good to me ... @roccotripaldi , can you take a look and see if works well for you too?

roccotripaldi commented:
doing some more testing now. i just tested with a non a8c account with less than 10 sites. and this branch works fine. you never even see the notice, because the site's list is updated quick. i'm going to try an other round with my a8c account.

roccotripaldi commented:
ok, it seems pretty close, but unfortunately i found an other piece of weirdness.
- connect Jetpack
- go to calypso.localhost:3000/plugins/$site
- notice that you are redirected to calypso.localhost:3000/plugins/
- go to calypso.localhost:3000/plugins/$site again
- this time you'll get the 'Waiting for site' flow

This screen cast demonstrates:
https://cloudup.com/csmpd-q0xGw

Agh. I'm not sure why it is happening.

ryelle commented:
I can't seem to replicate that, each time I try it correctly waits. Maybe your user site count isn't updated fast enough too? Are you sandboxing wp.com while testing this? (just since sandboxes tend to have slower response time)

roccotripaldi commented:
> Are you sandboxing wp.com while testing this?

Nope. Since it's working smoothly for both you and Javi, I'm tempted to say ship it!

ryelle commented:
Yay! Before we do that, @rickybanister can you chime in on what the notice should say? Currently it's "Waiting for site…". You can see the process starting at 0:58 in on Rocco's screencast: https://cloudup.com/csmpd-q0xGw

rickybanister commented:
How about 'Finishing set up...'

I sort of wish we could show the blinking dot in the masterbar or something while you wait—something about an animation make it feel like it's still working and hasn't stalled...is that possible?

ryelle commented:
I think for most people it's not long enough to seem stalled. I've updated the text, and I think if this is OK we should just push it rather than tweaking more — if we decide it does need an animation we can work on that in another issue/PR.

johnHackworth commented:
Yup, this is a pretty huge improvement from the current state of production, so let's ship this and if we need to iterate on it, let's do it on another branch! :ship:

aduth commented:
An undesirable consequence of this pull request (specifically c4fda7298e232ee0cd6c17d2b10f750dd4e7ee45) is that there are routes like the editor that render a shell of the page while the sites list is loading. With these changes, nothing is rendered until the sites is populated (especially noticeable on an empty `localStorage`). We should implement route controllers accounting for the fact that sites may or may not exist and render appropriately regardless.

mtias commented:
Since this behaviour is so specific to the connect flow, instead of affecting all the app blindly, I think we should be triggering the wait only if a query-string is present or something like that.

mtias commented:
I'm not sure I understand — there are many places in WP.com that link to Calypso specific areas without an issue. The problem here was when a new site was being connected for Jetpack that an existing sites list would not have it, so the app would infer the site didn't exist and redirect you to site selection.

johnHackworth commented:
>— there are many places in WP.com that link to Calypso specific areas without an issue. 

Yup, the problem is that those are wp-com sites... if you are in wp.com already, you will have that site on your sites list. But when you are in jetpack, and maybe you have even never ever loaded calypso, and then you click a (new) link that redirects you there... your site, even if it's connected to your acount, is not in your local sites list yet

mtias commented:
> if you are in wp.com already, you will have that site on your sites list

Not necessarily, only if I've visited a Calypso area before.

> and maybe you have even never ever loaded calypso, and then you click a (new) link that redirects you there

In this case you won't have sites-list data anyways.

roccotripaldi commented:
> The problem here was when a new site was being connected for Jetpack that an existing sites list would not have it, so the app would infer the site didn't exist and redirect you to site selection.

Now that all Jetpack connections are going through Calypso, and the site's list is loaded / refreshed during connection, this should be less of a problem. However, I don't have any data to back this up. Are there any stats we can look at?

nb commented:
Hey, I just experienced this, waiting on a blank page for 10-15 seconds when loading the posts page for a site. Is there a plan to fix it? It seems to impact a lot of areas of Calypso.

johnHackworth commented:
@ryelle  can you take a look at this? Maybe we could just remove this change, now that most of the jetpack sites are connected through JPC, but I'm not 100% sure about it

ryelle commented:
@johnHackworth Hmm, I thought like @roccotripaldi that this would be fixed now that everyone goes through Calypso. But... if this were really fixed no one would see the wait message. @nb Maybe you hit a different bug that this triggers this waiting code? Did you just connect a Jetpack site?

We didn't add any tracking about how long/often people end up waiting, but I'm happy to drop something in now, or I can just remove the wait time and see what happens - IIRC this happens to automatticians more than customers.

aduth commented:
The issue is due to the specific changes in c4fda7298e232ee0cd6c17d2b10f750dd4e7ee45 where we now wait until sites have been loaded to proceed to the next middleware in Page.js . We should not need to wait for sites loaded to continue through to the next middleware, otherwise nothing will be rendered.

mtias commented:
This affects every single My Sites page, you can reproduce by running `localStorage.clear()` and reloading the page. You will be forced to wait at the point @aduth mentions until sites-list finishes fetching.

roccotripaldi commented:
@ryelle 

>  this would be fixed now that everyone goes through Calypso

Yep, Jetpack Sites should no longer need to wait, now that every one is connecting via calypso. 


> But... if this were really fixed no one would see the wait message

I freshly connected Jetpack sites should no longer trigger the wait message. But sadly, other areas are, as a side-effect of our work, being affected.

So, i think we can safely remove the wait times that we added in c4fda7298e232ee0cd6c17d2b10f750dd4e7ee45

ryelle commented:
Added a PR to remove the wait, #9057

apeatling commented:
This is a good example where the change back could have happened much sooner. There's almost two weeks between flagging this as an issue, and a PR to remove it. What can we do to speed up this process?

Some suggestions:

* Get on a zoom chat and make a decision about what to do
* Someone make a decision and add a PR for removal, so discussion is more contained

Also as an aside, if a change is going in that requires a user to wait, we should think long and hard about whether there is an alternative. We want to make Calypso fast and responsive, removing the impact of updates blocking UI progress.
